### PR TITLE
[python] First step in supporting type inference, based on the usage of `sq_concat`

### PR DESCRIPTION
**Pull request by jefremof**

### CODE REVIEW

File: usvm-python/usvm-python-main/src/main/kotlin/org/usvm/machine/interpreters/symbolic/operations/basic/MethodNotifications.kt [hunk 1]
```diff
@@ -37,7 +38,10 @@ fun nbAddKt(
     context.ctx
 ) {
     context.curState ?: return
-    pyAssert(context, left.evalIsSoft(context, HasNbAdd) or right.evalIsSoft(context, HasNbAdd))
+    val nbAdd = left.evalIsSoft(context, HasNbAdd) or right.evalIsSoft(context, HasNbAdd)
+    val sqConcat = left.evalIsSoft(context, HasSqConcat) and right.evalIsSoft(context, HasSqConcat)
+    pyAssert(context, context.ctx.mkImplies(nbAdd.not(), sqConcat))
```
Comment (tochilinak) [inline usvm-python/usvm-python-main/src/main/kotlin/org/usvm/machine/interpreters/symbolic/operations/basic/MethodNotifications.kt:43]: `context.ctx.mkImplies(nbAdd.not(), sqConcat)` -> `nbAdd.not() implies sqConcat`

Comment (tochilinak) [reply usvm-python/usvm-python-main/src/main/kotlin/org/usvm/machine/interpreters/symbolic/operations/basic/MethodNotifications.kt:43]: Please add comment that describes the logic behind these changes.

Comment (tochilinak) [reply usvm-python/usvm-python-main/src/main/kotlin/org/usvm/machine/interpreters/symbolic/operations/basic/MethodNotifications.kt:43]: Also, please add the method that catches the event of `sqConcat` (similar to `nbAdd`). In that method, add constraint that operands have `sq_concat` method. In some cases, execution might skip `nbAdd` and go straight to `sqConcat`, but we still need to fix that operands have this method.

**Review by tochilinak — CHANGES_REQUESTED** [review]

File: usvm-python/usvm-python-main/src/main/kotlin/org/usvm/machine/interpreters/symbolic/operations/basic/MethodNotifications.kt [hunk 2]
```diff
@@ -37,7 +38,15 @@ fun nbAddKt(
     context.ctx
 ) {
     context.curState ?: return
-    pyAssert(context, left.evalIsSoft(context, HasNbAdd) or right.evalIsSoft(context, HasNbAdd))
+    /*
+    The __add__ method corresponds both to the nb_add and sq_concat slots,
+    so it is crucial not to assert the presence of nb_add, but to fork on these
+    two possible options.
```
Comment (tochilinak) [inline usvm-python/usvm-python-main/src/main/kotlin/org/usvm/machine/interpreters/symbolic/operations/basic/MethodNotifications.kt:44]: Also, please add the following:

```
Moreover, for now it was decided, that operation `sq_concat` makes sense only in the situation, when both operands have the corresponding slot.
```

**Review by tochilinak — APPROVED** [review]

**Review by sergeypospelov — APPROVED** [review]

**Pull request merged**